cwms_radar_api/src/main/java/cwms/radar/security/CwmsAccessManager.java

```diff
@@ -0,0 +1,204 @@
+import io.javalin.core.security.RouteRole;
+import io.javalin.http.Context;
+import io.javalin.http.Handler;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jooq.DSLContext;
+import org.jooq.SQLDialect;
+import org.jooq.impl.DSL;
+
+import usace.cwms.db.jooq.codegen.packages.CWMS_ENV_PACKAGE;
+
+public class CwmsAccessManager implements AccessManager
+{
+	public static final Logger logger = Logger.getLogger(CwmsAccessManager.class.getName());
+	public static final String DATABASE = "database";
+
+	@Override
+	public void manage(@NotNull Handler handler, @NotNull Context ctx, @NotNull Set<RouteRole> requiredRoles)
+			throws Exception
+	{
+		boolean shouldProceed = isAuthorized(ctx, requiredRoles);
+
+		if(shouldProceed)
+		{
+			setSessionKey(ctx);
+
+			// Let the handler handle the request.
+			handler.handle(ctx);
+		}
+		else
+		{
+			String msg = getFailMessage(ctx, requiredRoles);
+			logger.info(msg);
+			ctx.status(HttpServletResponse.SC_UNAUTHORIZED).json(RadarError.notAuthorized());
+			ctx.status(401).result("Unauthorized");
+		}
+
+	}
+
+	private void setSessionKey(@NotNull Context ctx)
+	{
+		String sessionKey = getSessionKey(ctx);
+		if(sessionKey != null)
+		{
+			// Set the user's session key in the database.
+			Connection conn = ctx.attribute(DATABASE);
+			setSession(conn, sessionKey);
+		}
+	}
+
+	@NotNull
+	private String getFailMessage(@NotNull Context ctx, @NotNull Set<RouteRole> requiredRoles)
+	{
+		Set<RouteRole> specifiedRoles = getRoles(ctx);
+		Set<RouteRole> missing = new LinkedHashSet<>(requiredRoles);
+		missing.removeAll(specifiedRoles);
+
+		return "Request for: " + ctx.req.getRequestURI() + " denied. Missing roles: " + missing;
+	}
+
+
+
+	@Nullable
+	private Principal getPrincipal(@NotNull Context ctx)
+	{
+		//		CwmsUserPrincipal principal = ctx.attribute("principal"); // wrong
+		Principal userPrincipal = ctx.req.getUserPrincipal();
+
+		return userPrincipal;
+	}
+
+	public boolean isAuthorized(Context ctx, Set<RouteRole> requiredRoles)
+	{
+		boolean retval;
+		if(requiredRoles == null || requiredRoles.isEmpty())
+		{
+			retval = true;
+		}
+		else
+		{
+			Set<RouteRole> specifiedRoles = getRoles(ctx);
+			retval = specifiedRoles.containsAll(requiredRoles);
+		}
+		return retval;
+	}
+
+	public Set<RouteRole> getRoles(Context ctx){
+		Set<RouteRole> retval = new LinkedHashSet<>();
+		if(ctx != null)
+		{
+			Principal principal = getPrincipal(ctx);
+			Set<RouteRole> specifiedRoles = getRoles(principal);
+			if(!specifiedRoles.isEmpty())
+			{
+				retval.addAll(specifiedRoles);
+			}
+		}
+		return retval;
+	}
+
+	@Nullable
+	private String getSessionKey(@NotNull Context ctx)
+	{
+		String sessionKey = null;
+
+		Principal principal = getPrincipal(ctx);
+
+		if(principal != null){
+			sessionKey = callGetSessionKeyReflectively(principal);
+		}
+		return sessionKey;
+	}
+
+	private String callGetSessionKeyReflectively( Principal principal)
```

Comment:
I feel like I've failed somewhere, why not cast the Principal to the `CwmsUserPrincipal`?

Reply:
or just totally missunderstood part of how java worked. Either way I'd like to know what led to the reflection.

Reply:
I also did not expect this.  I believe its a classloader issue.   If I look at the object with a debugger in a test case it believes its a TestCwmsUserPrincipal which extends CwmsUserPrincipal but it throws an exception if I try to cast it to CwmsUserPrincipal.  If the object was created by cwms_aaa it must have a different classloader than radar b/c we could deploy/redeploy then separately.

Reply:
hmm. That, unfortunately, makes a lot of sense. the CWMS_AAA code would be in the system classloader and locally here in the app class loader. 

The fix is likely switching this to compile only/provide scope and making sure the api jar is in the system path. I will investigate later. We'll leave it for now.